rfcs/0158-artifact-metadata.md

```diff
@@ -0,0 +1,62 @@
+# RFC 158 - artifact metadata
+* Comments: [#158](https://github.com/taskcluster/taskcluster-rfcs/pull/158)
+* Proposed by: @escapewindow
+
+# Summary
+
+The goal is to provide Artifact Integrity guarantees from the point the worker uploads an artifact, to the point that someone downloads that artifact to use. We can do this by:
+
+1. adding SHA metadata to artifacts in the Queue,
+2. ensuring that that metadata can't be modified once it's written, and
+3. providing a download tool that queries the Queue for an artifact's location and SHA, downloads the artifact, and verifies the downloaded SHA matches the SHA provided by the Queue.
+
+## Motivation
+
+This would improve robustness and security. By storing a SHA to verify on download, we can avoid corrupt downloads. By verifying the read-only SHA before we use the artifact, we can detect malicious tampering of artifacts at rest before we use it for a critical operation.
+
+(This would obsolete the [artifacts section of the Chain of Trust artifact](https://scriptworker.readthedocs.io/en/latest/chain_of_trust.html#chain-of-trust-artifacts), and improve artifact integrity platform-wide.)
+
+# Details
+
+## Adding artifact metadata to the Queue
+
+First, we add a `metadata` dictionary to the `S3ArtifactRequest` type. This is a dictionary to allow for flexibility of usage. The initial known keys would include
+
+```
+ContentFilesize int64 `json:"contentFilesize"`
+ContentSha256 string `json:"contentSha256"`
+ContentSha512 string `json:"contentSha512"`
+```
+
+The sha256 field is required for Artifact Integrity. Mozilla Releng has use cases for all 3 fields, so I'm proposing all 3.
+
+A future entry may be `ContentSha256WorkerSignature`, once we solve worker identity.
+
+We can optionally add a `metadata` dictionary to the `ErrorArtifactRequest` and `RedirectArtifactRequest` types.
+
+A new `Queue.getArtifactInfo` endpoint will return the artifact URL and metadata.
+
+## Ensuring that metadata can't be modified once it's written
+
+
+* Queue would have no API to write to this data
+* The backend storage (postgres) would be configured to not allow updates that would modify the data (so only select, insert, delete, not update)
+
+The latter point depends on the postgres migration.
+
+## Providing a download tool
+
+This tool queries the Queue for the artifact metadata, downloads the artifact, and verifies any shas. We should allow for optional and required metadata fields, and for failing out if any required information is missing, or if the sha doesn't match. We should be sure to measure the shas and filesizes on the right artifact state (e.g. combining a multipart artifact, not compressed unless the original artifact was compressed).
+
+This tool should be usable as a commandline tool, or as a library that the workers can use.
```

Comment:
Will this also be included in Taskcluster Users workflows? i.e. QA downloading an artifact to test?

If so, how will that be "enforced" or "promoted"?

Reply:
I think we would start with encouraging and recommending the tool, and making it easy to get and use. Once that starts happening, we could update user docs to mention the tool, and remove references to other ways to download things. But our main use case here is protecting the release pipeline: making sure no malicious artifact, worker, or task can result in us shipping something arbitrary. If we end up not enforcing manual use by humans, but those manual tasks never result in our shipping something we didn't expect to ship, we may be ok with that level of risk.

Reply:
Ok, sounds good.